File: Hailstone-sequence/Swift/hailstone-sequence.swift
func hailstone( m:Int) -> [Int] {
	var n=m;
    var arr = [n]

    while n != 1 {

        if n % 2 == 0 {
            n /= 2
        } else {
            n = (3 * n) + 1
        }

        arr.append(n)
    }

    return arr
}

let n = hailstone(m:27)

print("hailstone(27): \(n[0...3]) ... \(n[n.count-4...n.count-1]) for a count of \(n.count).")

var longest = (n: 1, len: 1)

for i in 1...100_000 {

    let new = hailstone(m:i)

    if new.count > longest.len {
        longest = (i, new.count)
    }
}

print("Longest sequence for numbers under 100,000 is with \(longest.n). Which has \(longest.len) items.")

File: Hailstone-sequence/Dart/hailstone-sequence.dart
import 'dart:collection';

List<int> hailstone(int n) {
  if(n<=0) {
    throw new IllegalArgumentException("start value must be >=1)");
  }
  Queue<int> seq=new Queue<int>();
  seq.add(n);
  while(n!=1) {
    n=n%2==0?(n/2).toInt():3*n+1;
    seq.add(n);
  }
  return new List<int>.from(seq);
}

// apparently List is missing toString()
String iterableToString(Iterable seq) {
  String str="[";
  Iterator i=seq.iterator();
  while(i.hasNext()) {
    str+=i.next();
    if(i.hasNext()) {
      str+=",";
    }
  }
  return str+"]";
}

main() {
  for(int i=1;i<=10;i++) {
    print("h($i)="+hailstone(i).join());
  }
  List<int> h27=hailstone(27);
  List<int> first4=h27.getRange(0,4);
  print("first 4 elements of h(27): "+first4.join(" "));

  List<int> last4=h27.getRange(h27.length-4,h27.length);
  print("last 4 elements of h(27): "+last4.join(" "));

  print("length of sequence h(27): "+h27.length.toString());

  int seq,max=0;
  for(int i=1;i<=100000;i++) {
    List<int> h=hailstone(i);
    if(h.length>max) {
      max=h.length;
      seq=i;
    }
  }
  print("up to 100000 the sequence h($seq) has the largest length ($max)");
}
